EmptyConstraint - allow null string

jhamm opened the pull request:
fix #1171 

The cases `Assert.That(null, Is.Not.Empty)` and `Is.Not.Null.Or.Empty` will continue to fail as they are written. The 'null' value cannot be checked to see if the EmptyConstraint types can be applied because this is lost.

If the test were rewritten as:
````
string actual = null;
Assert.That(actual, Is.Empty);

//or

Assert.That((string)null, Is.Empty);
````

I think that most tests will have the type information like the rewritten test cases as opposed to having just `null`.

These tests throw the AssertionException as expected. If the type cannot be applied, (not string, collection, or DirectoryInfo), the ArgumentException will be thrown.

CharliePoole commented on src/NUnitFramework/framework/Constraints/EmptyConstraint.cs:
```diff
@@ -48,19 +48,25 @@ public override string Description
+            var actualType = GetBaseType(actual);
 
-            if (actual is string)
+            if (actual is string || actualType == typeof(string))
```
I don't understand the added term. Can you give an example where it is neede?

CharliePoole commented:
It seems like some new EmptyConstraint tests are needed. I realize that your Assert.That tests test the change indirectly, but we generally try to have direct tests for each constraint. See file EmptyConstraintTests.cs.

jhamm commented:
Thanks, I agree that direct tests are a good idea. I have added two tests to EmptyConstraintTests.

The previous check `actual is string` will not work if `actual` is null. The change to use `actualType == typeof(string)` will evaluate to true even if `actual` is null.

I have also moved the throw ArgumentException to immediately after the string check.

rprouse commented:
Looks good. @CharliePoole can you take a second look?

CharliePoole commented:
We started a discussion of how it should work on the issue, so maybe we need to hold this until that is completed and then check to see if it actually does what we decide the constraint should do.

jhamm commented:
Based on the discussion in #1171, this can be closed and not merged.

CharliePoole commented:
In that case, we should at least check to see if our current behavior matches what we defined as desirable in the discussion.
